### DISCUSSION TITLE
Cannot use Threads in windows and android

### DISCUSSION THREAD

**Discussion by Spitskool42** [beeware/toga#1538] [Q&A]
**Describe the bug**
Cannot use Threads in windows and android.
I was able to find a solution for windows by using `loop.run_in_executor` to add a thread on to the current event loop. But this did not work on android :(

**To Reproduce**
Run example program on windows or android.

**Expected behavior**
I was expecting that we could use a different thread to execute a function that is blocking. A function I use can take up to 3 seconds.
Only found 1 solution for in windows, but does not work on android.

**Screenshots**
See example code below.

**Environment:**
 - Operating System: Android 11 and windows 10
 - Python version: 3.10.5
 - Software versions:
   - Briefcase: 0.3.7
   - Toga: 0.3.0 I think
   - ...

**Additional context**
Example program with 2 ways to call blocking functions below
First button will create a Thread that blocks the whole GUI in windows an android for 5 seconds.
Second button will add a function to the main event loop, this will not block in windows, but will in android.
``` python
"""
My first application
"""
import asyncio
import time
from threading import Thread
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW


class Testblocking(toga.App):

    def fThread1(self):
        i=1
        for x in range(5):
            time.sleep(1)
            self.Text.text = f'Test 1: {i}'
            i+=1

    def fButtonclick(self, widget):
        self.Thread1 = Thread(target=self.fThread1())

    def fasyncioExecutor(self):
        i=1
        for x in range(5):
            time.sleep(1)
            self.Text2.text = f'Test 2: {i}'
            i+=1
        self.buttonB.enabled=True

    async def fButtonclick2(self, widget):
        self.buttonB.enabled=False
        loop = asyncio.get_running_loop()
        await loop.run_in_executor(None, self.fasyncioExecutor)

    def startup(self):
        main_box = toga.Box(style=Pack(direction=COLUMN))

        buttonA = toga.Button('testBlocking 1 new thread', on_press=self.fButtonclick)
        self.Text = toga.Label('Test 1')
        main_box.add(buttonA, self.Text)

        self.buttonB = toga.Button('testBlocking 2 new asyncion', on_press=self.fButtonclick2)
        self.Text2 = toga.Label('Test 2')
        main_box.add(self.buttonB, self.Text2)

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()

def main():
    return Testblocking()

```

**Comment by freakboy3742** [comment]
Ok - so there's no bug here. This is an error of usage, on multiple counts.

### Raw threads

Button 1 is using `Thread(target=self.fThread1())`. This means `self.fThread1()` is being invoked, and the return value of that function (`None`) is being used as the target of a `Thread` instance. The thread is then unused, because it isn't ever started. This also means that the GUI event loop is being directly blocked by the calls to sleep(), which means the app freezes until control is released to the event loop - when `fThread1()` completes.

If you modify `fButtonclick()` to:
```
self.Thread1 = Thread(target=self.fthread1)
self.Thread1.start()
```

the thread will run in the background. You're still going to hit problems on Windows, though, because Windows doesn't permit you to perform GUI operations on threads other than the main thread. You can prove the thread is working by using `print()` calls rather than setting label text; but if you update the label text, your app will do something between locking up and crashing, depending on which way the wind is blowing. 

There are two ways to fix this.

Option 1 is... "don't do that" :-) and make your background thread a pure "worker". If you've got something like an internal web server, this is a good option. The worker thread only serves web requests; any GUI operation is done on the main thread contacting the web server thread. Another use case: if you've got an calculation engine in the background, it can happily compute values, and modify values on the app instance; the app then have an asynchronous "update GUI" loop that polls for the current value and updates/redraws the GUI as needed.

Option 2 is to tell the thread to post a message to the main GUI loop. You do this using `app.add_background_task()`. Instead of calling `self.Text1.text = "..."` directly in the thread logic, you use:
```
async def update_gui(app):
    app.Text1.text = f'loop {i}'
self.add_background_task(update_gui)
```

This creates a method that will make the desired GUI change, and posts that message to the app's main event loop. It will be picked up at the earliest opportunity, and the GUI will be updated.

### Async "thread"

The async "thread" implementation has a similar problem. Although you're using an async event handler, and posting a method to the executor, what you're posting is a synchronous method. This means you lose all the benefits of being async, because the blocking API calls (in this case, `time.sleep()`) stop the asyncio event loop. 

You can avoid this (and dramatically simplify the process) by using a non-blocking sleep implementation: `asyncio.sleep()`
```
    async def fButtonclick2(self, widget):
        self.buttonB.enabled=False
        for i in range(5):
            await asyncio.sleep(1)
            self.Text2.text = f'Test 2: {i}'
            i+=1
        self.buttonB.enabled=True
```

There's no need to use a separate method or post something to the executor - you create a method that is non-blocking, using non-blocking primitives. The button handler then has permission to hand control back to the GUI thread whenever it hits the sleep call, which means the GUI can redraw.

**Comment by Spitskool42** [comment]
Thank you for the good detailed explanation @freakboy3742 !
My main mistake was that I used `Thread(target=self.fthread1 ( ) )` instead of `Thread(target=self.fthread1)`, This will immediately start the function. Never used threads in python before :(

I now changed the program with 2 solutions you mentioned that work on windows! 
1 method uses asyncio, and the other uses a thread plus add_background_task method from toga to update GUI.

The solution I need is the one that creates his own thread. I cannot convert my apps method to asyncio.
But this 'thread' method does not work on android, It will crash the app on android after I click the first button and wait a while or click the second button. And it also does not update the GUI from text1, only once before the crash. Asyncio method works fine.

And the android app crashed without any info in the debug window. Maybe android does not allow creating of different threads like the way python (threading) does?


### Program that works on windows correctly:
``` python
"""
My first application
"""
import asyncio
import time
from threading import Thread
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW


class Testblocking(toga.App):

    def fThread1(self):
        self.Text1_i=1
        for x in range(5):
            time.sleep(1)
            self.app.add_background_task(self.update_gui)
            self.Text1_i+=1

    async def update_gui(self, widget):
        #This updates the Text in the GUI, is slightly delayed (waiting to be executed by event loop), that is why it counts from 2 to 6.
        self.Text1.text = f'Test 1: {self.Text1_i}'

    def fButtonclick(self, widget):
        self.Thread1 = Thread(target=self.fThread1)
        self.Thread1.start()

    async def fasyncioExecutor(self):
        self.buttonB.enabled=False
        i=1
        for x in range(5):
            await asyncio.sleep(1)
            self.Text2.text = f'Test 2: {i}'
            i+=1
        self.buttonB.enabled=True

    async def fButtonclick2(self, widget):
        await self.fasyncioExecutor()

    def startup(self):
        main_box = toga.Box(style=Pack(direction=COLUMN))
        buttonA = toga.Button('testBlocking 1 new thread', on_press=self.fButtonclick)
        self.Text1 = toga.Label('Test 1')
        main_box.add(buttonA, self.Text1)

        self.buttonB = toga.Button('testBlocking 2 new asyncion', on_press=self.fButtonclick2)
        self.Text2 = toga.Label('Test 2')
        main_box.add(self.buttonB, self.Text2)

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()

def main():
    return Testblocking()

```

**Reply by Spitskool42** [reply]
The code `self.app.add_background_task(self.update_gui)` seems to be crashing the android app.

**Reply by Spitskool42** [reply]
Oke, now I created one blocking thread that also works on Android. But it uses some kind of polling to update the GUI, Not really ideal. It uses await asyncio.sleep(1) (like in the handler example) to update the text every second if changed. It works on windows and android.

Maybe the add_background_task is not working like intended on android? 
<s>It seems that I only can use the Asyncio background task. On windows the background_task with normal method works.</s>
The `app.add_background_task()` is allowed inside a different thread in windows, but crashes in android.

### The example program that works on both:
``` python
import asyncio
import time
from threading import Thread
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW


class Testblocking(toga.App):

    def fThread1(self):
        self.Text1_i=1
        for x in range(5):
            time.sleep(1)
            print(self.Text1_i)
            self.Text1_i+=1  
  
    async def update_gui(self, widget, **kwargs):
        i=0
        while True:
            if self.Text1.text != f'Test 1: {self.Text1_i} update count: {i}':
                i+=1 
                self.Text1.text = f'Test 1: {self.Text1_i} update count: {i}'
                self.Text1.refresh()
            if not self.Thread1.is_alive():
                self.buttonA.enabled=True   
            await asyncio.sleep(1)

    def fButtonclick(self, widget):
        self.buttonA.enabled=False
        self.Thread1 = Thread(target=self.fThread1)
        self.Thread1.start()

    async def fasyncioExecutor(self):
        self.buttonB.enabled=False
        i=1
        for x in range(5):
            await asyncio.sleep(1)
            self.Text2.text = f'Test 2: {i}'
            i+=1
        self.buttonB.enabled=True

    async def fButtonclick2(self, widget):
        await self.fasyncioExecutor()

    def startup(self):
        main_box = toga.Box(style=Pack(direction=COLUMN))
        self.buttonA = toga.Button('testBlocking 1 new thread', on_press=self.fButtonclick)
        self.Text1 = toga.Label('Test 1: 0 update count: 0')
        main_box.add(self.buttonA, self.Text1)

        self.buttonB = toga.Button('testBlocking 2 new asyncion', on_press=self.fButtonclick2)
        self.Text2 = toga.Label('Test 2')
        main_box.add(self.buttonB, self.Text2)

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()

        self.Text1_i=0
        self.Thread1 = Thread()
        self.add_background_task(self.update_gui)

def main():
    return Testblocking()

```

**Reply by Spitskool42** [reply]
Sorry for this many replies. But I found out that without this loop below the android threads that are started do not work (no info in the debugger console, just nothing from the thread seems to be executed):

```python
async def update_gui(self, widget, **kwargs):
    while True:
        await asyncio.sleep(1)

def startup(self):
     self.add_background_task(self.update_gui)
```
So I just added this code to my app, and my own thread works. Any idea why this is needed?

**Reply by Spitskool42** [reply]
> Sorry for this many replies. But I found out that without this loop below the android threads that are started do not work (no info in the debugger console, just nothing from the thread seems to be executed):
> 
> ```python
> async def update_gui(self, widget, **kwargs):
>     while True:
>         await asyncio.sleep(1)
> 
> def startup(self):
>      self.add_background_task(self.update_gui)
> ```
> 
> So I just added this code to my app, and my own thread works. Any idea why this is needed?

This method for android seems to not work correctly.
the longer the app stays open, the more Native invocation's are pushed to the debug windows
Seems like every asyncio.sleep(1) (1 second) a new native invocation is added to the existing one:
![image](https://user-images.githubusercontent.com/109864670/182038771-b7740855-de53-4102-ae9f-998dc2a3a705.png)
In the application itself you can also notice a small freeze after a while when you for example are scrolling down longer than 1 second.

@freakboy3742 
This also means that the Toga example: handlers is working, but not correctly on android. Seems like something is going wrong.

**Reply by freakboy3742** [reply]
Some of these problems I can't reproduce; and some I can't work out what you think the problem is.

The log messages you describe are completely expected - they're event handlers reporting "I'm about to start running Python code" and "I've finished running Python code". 

I've just run the handlers demo on Android (on the emulator), and I can't see any problems. If you think it's not working, you'll need to provide specifics of the behavior you're seeing (or not seeing) that you think you should (or shouldn't) be seeing.

The last example code you've provided doesn't do anything... because you haven't asked it to do anything. It's literally a `while True: sleep(1)` loop. It doesn't *do* anything. I'm not clear on what it is you think it *should* be doing.

The problems I *can* reproduce are all related to threads. However, in most cases, it's not as simple as "not working". 

Threads are definitely starting and running; your last full example demonstrates a fully working example of threads.

The the earlier example is failing for 2 reasons - one is actually a failure, and one is a red herring.

The red herring is that stdout/err isn't playing well with threads. stdout is only reflected in the log until the point where then button handler returns. The Thread is still running, and output is still being generated, but it isn't being reliably redirected to the log. This isn't *entirely* surprising; the log redirection is a big of a hack; my guess is that there's some garbage collection going on with a stdout handler. I definitely won't call this working 100% correctly, but the thread is definitely *running*.

The piece that is definitely breaking is adding a background task while *in* a thread. I'm not sure why this is breaking; my guess is that there is something being garbage collected by Java as it moves between threads, which means something needs to be made a global JNI reference. However, at least as a short term workaround, your "working" example works - and the approach to polling for output (a single long-lived polling worker, started from the main thread) is the approach I would generally recommend.

**Reply by freakboy3742** [reply]
I've just added #1539 as a record of the background worker problem.

**Reply by Spitskool42** [reply]
> Some of these problems I can't reproduce; and some I can't work out what you think the problem is.
> 
> The log messages you describe are completely expected - they're event handlers reporting "I'm about to start running Python code" and "I've finished running Python code".
> 
> I've just run the handlers demo on Android (on the emulator), and I can't see any problems. If you think it's not working, you'll need to provide specifics of the behavior you're seeing (or not seeing) that you think you should (or shouldn't) be seeing.
> 
> The last example code you've provided doesn't do anything... because you haven't asked it to do anything. It's literally a `while True: sleep(1)` loop. It doesn't _do_ anything. I'm not clear on what it is you think it _should_ be doing.
> 

Thank you for the reply, I will try to explain it a bit more detailed. The background task from the handlers demo (our my example with only `asyncio.sleep(1)` in a while loop), seems to increase it's number of native invocations every second. 

The Native invocation triggered by our background task every second starts like this (look a the time stamp seconds):
![image](https://user-images.githubusercontent.com/109864670/182122001-b23ac2f8-875d-459c-86ee-35817d90c662.png)

But ends like this, and no user input was given at all:
![image](https://user-images.githubusercontent.com/109864670/182122138-4fe585ea-1fdc-4953-869c-508bf42c047b.png)

Maybe this is because it works like an even handler that subscribes to the function every second another time? And that in the end so many calls per second are executed that the app gets a lag spike/freeze every second.
I have also seen this in my App, when it is running for a longer period of time, the scrolling or opening the menu motions are getting a freeze or small lag spike every second (Not really annoying but it is noticeable).

**Reply by freakboy3742** [reply]
Ah, that makes more sense, and yes, that definitely looks like something is going wrong - I'm going to guess that expended handlers aren't being fully cleaned up when they're done.

**Reply by mhsmith** [reply]
> The red herring is that stdout/err isn't playing well with threads. stdout is only reflected in the log until the point where then button handler returns. The Thread is still running, and output is still being generated, but it isn't being reliably redirected to the log. This isn't _entirely_ surprising; the log redirection is a big of a hack; my guess is that there's some garbage collection going on with a stdout handler. I definitely won't call this working 100% correctly, but the thread is definitely _running_.

I wasn't able to reproduce this with the code above. For me, the 1, 2, 3, 4, 5 prints all came through with level `I` and tag `Python`, mixed in with the `D Python` "Native invocation" messages shown above.

I was confused about the tag, until I found that Rubicon actually does its own redirection of the Python stdout/stderr objects. So the C-level redirection in the Android template's native-lib.cpp (with tag `stdio`) will only affect the C-level streams used by non-Python code.

@freakboy3742: If you can still reproduce this, can you report it as a separate issue with a minimal test case?

**Reply by Spitskool42** [reply]
@mhsmith I think what @freakboy3742  means is that print() in a new thread is not logged in the console. It will show the values after some other ansycio or main thread is logged.

You can replicate it with the program I wrote under 'The example program that works on both:'.
And removing the last line in def startup(self): `#self.add_background_task(self.update_gui)`. Removing the asyncio poll that await sleeps every second.
Now the print() values from the thread are not printed. Only when pushing the second button, and shows the old print() values.

**Reply by freakboy3742** [reply]
@mhsmith I think I've narrowed the source of the problem. It's a doozy; see #1546 for a full description and reproduction case.

**Reply by mhsmith** [reply]
And see #1542 for the handler cleanup issue.

It's annoying that GitHub doesn't automatically post reverse links when an issue is mentioned in a discussion or vice versa. Personally, I would prefer not to use discussions at all until this is fixed.

**Reply by freakboy3742** [reply]
Agreed it's an annoying oversight/ommission - but any other discussion forum or mailing list tool wouldn't give us 2-way Github integration, so I'm not convinced that alone is a reason to abandon Discussions as a way to have long-form persistent threaded conversations that aren't directly related to an issue.

**Comment by Spitskool42** [comment]
Oke, I maybe found out a new related way why android crashes:

### To reproduce:
Run the program below on android. This works fine, the first button creates a new .txt file after 1 second, and button 2 reads the text file that was created.

But when you exchange line 21 by line 22, `fCurrentNote = open(self.filepath + '/test1.txt', 'w')` by `fCurrentNote = open(str(self.app.paths.data) + '/test1.txt', 'w')`, the android app crashes when first clicking button1 and then button2 to read the created file. I think when calling a function or class in a new thread that is related to the main thread it cannot find these references? When reopening the app after it is crashed and the text file still has the old value saved, was not updated, but the console printed the new value.

``` python
"""
My first application
"""
import os
import pdb
from random import randint
import time
from threading import Thread
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW


class Testblocking2(toga.App):

    def fThread1(self):   
        if self.Text1_i==1:
            time.sleep(1)
            self.Text1_i = randint(2, 128)
            print(self.Text1_i)
            fCurrentNote = open(self.filepath + '/test1.txt', 'w')
            #fCurrentNote = open(str(self.app.paths.data) + '/test1.txt', 'w')
            fCurrentNote.write('-------- ' + str(self.Text1_i))  
            fCurrentNote.close()
            self.Text1_i+=1 

    def fButtonclick(self, widget):
        self.Text1_i=1 
        self.Thread1 = Thread(target=self.fThread1)
        self.Thread1.start()

    async def fButtonclick2(self, widget):
        try:
            fCurrentNote = open(self.filepath + '/test1.txt', 'r')
            self.Text1.value = 'Test 1 file reads: ' + fCurrentNote.read()
            fCurrentNote.close()
            #os.remove(self.filepath + '/test1.txt')
        except Exception as e:
            self.Text1.value = 'Test 1 file reads: ' + str(e)

    def startup(self):
        main_box = toga.Box(style=Pack(direction=COLUMN))
        self.buttonA = toga.Button('testBlocking 1 Create random value in thread', on_press=self.fButtonclick)
        self.Text1 = toga.TextInput(initial='Test 1: ')
        main_box.add(self.buttonA, self.Text1)

        self.buttonB = toga.Button('testBlocking 2 read file created by button 1', on_press=self.fButtonclick2)
        main_box.add(self.buttonB)

        self.filepath = str(self.app.paths.data)
        self.Text1_i=1
        if not os.path.exists(self.filepath):
            os.makedirs(self.filepath, 0o777)

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()

def main():
    return Testblocking2()

```

**Reply by freakboy3742** [reply]
Seems plausible - if the app object isn't "visible", and the thread is silently crashing, that would be consistent with the behavior you've described.
